how-it-works/3_chainevolution/02_fault-tolerance.subpage.md

```diff
@@ -7,7 +7,41 @@ slug: fault-tolerance
+
+Internet Computer follows a mix of both the approaches. The consensus protocol is divided into epochs. Each epoch is a few hundred consensus rounds. At the beginning of each epoch, all the nodes make a backup copy of their blockchain state, and create a “catch up package” (CUP). The CUP at height h contains all relevant information required for consensus to resume from height h. This includes hash of the blockchain state after processing block at height h. The CUP is then signed by at least 2/3rd of the nodes in the subnet. Each well-functioning node then broadcasts the CUP. 
+
+All the nodes in the subnet listen to the CUP messages broadcasted by their peers. Suppose a node observes that a received CUP has a valid signature (signed by at least 2/3 of the nodes in the subnet) and has a different blockchain state hash than the locally available state hash. Then the node initiates a  [state sync protocol](https://www.youtube.com/watch?v=WaNJINjGleg) to sync the blockchain state at that height (the height when the CUP is published). The blockchain state is organized as a merkle tree and could be a few hundreds of gigabytes. The syncing node might already have a majority of the blockchain state and need not download everything. Therefore, the syncing node tries to download only the subtrees of the peers’ blockchain state that differ from its local state. The syncing node first requests for the children of root of blockchain state. The syncing node then recursively downloads the subtrees that differ from its local state. 
+
+Note that while the failed/newly joined nodes are syncing the blockchain state, rest of the well-functioning nodes continue to process new blocks and make progress. The well-functioning nodes use their backup copy of the blockchain state (created at the same time as the CUP) to supply the state to syncing nodes. After the syncing node finishes syncing the blockchain state, it will request its peers for the consensus blocks generated in the meantime and process the blocks one by one. Once fully synced, the node can then process messages regularly as other nodes. 
+
+## Handling Regular Subnet Failures
+In rare cases, an entire subnet could get stuck and fail to make progress. A subnet could fail due to many reasons such as software bugs and non-deterministic execution. This could also happen when more than 1/3rd of the nodes in the subnet fail at the same time. In this case, the well-functioning nodes fail to create and sign a Catch Up Package (CUP), and thereby the failed nodes cannot recover automatically. 
```

Comment:
```suggestion
In rare cases, an entire subnet could get stuck and fail to make progress. A subnet could fail due to many reasons such as software bugs that lead to non-deterministic execution. This could also happen when more than 1/3rd of the nodes in the subnet fail at the same time. In this case, the well-functioning nodes fail to create and sign a catch-up package (CUP), and thereby the failed nodes cannot recover automatically. 
```